plugins/mongo_db_plugin/include/eosio/mongo_db_plugin/bson.hpp

```diff
@@ -0,0 +1,248 @@
+                  const fc::variant_object& obj = value.get_object();
+                  if (obj.size() == 2) {
+                     auto t = obj.begin();
+                     auto i = t;
+                     ++i;
+                     if (t->key().compare("t") == 0 && i->key().compare("i") == 0) {
+                        bsoncxx::types::b_timestamp ts;
+                        ts.timestamp = static_cast<uint32_t>(t->value().as_uint64());
+                        ts.increment = static_cast<uint32_t>(i->value().as_uint64());
+                        c.append(ts);
+                        break;
+                     }
+                  }
+               }
+            }
+         }
+         bsoncxx::builder::core sub(false);
+         to_bson(o, sub);
+         c.append(sub.extract_document());
+         return;
+      }
+      default:
+         FC_THROW_EXCEPTION(
+            fc::invalid_arg_exception,
+            "Unsupported fc::variant type: " + std::to_string(v.get_type()));
+   }
+}
+
+bsoncxx::document::value to_bson(const fc::variant& v)
+{
+   bsoncxx::builder::core doc(false);
+   if (v.get_type() == fc::variant::object_type) {
+      const fc::variant_object& o = v.get_object();
+      to_bson(o, doc);
+   }
+   else if (v.get_type() != fc::variant::null_type) {
+      FC_THROW_EXCEPTION(
+         fc::invalid_arg_exception,
+         "Unsupported root fc::variant type: " + std::to_string(v.get_type()));
+   }
+   return doc.extract_document();
+}
+
+void from_bson(const bsoncxx::document::view& view, fc::mutable_variant_object& o)
+{
+   for (bsoncxx::document::element ele : view) {
+      fc::variant v;
+      from_bson(ele, v);
+      o(ele.key().data(), v);
+   }
+}
+
+void from_bson(const bsoncxx::array::view& bson_array, fc::variants& a)
+{
+   a.reserve(std::distance(bson_array.cbegin(), bson_array.cend()));
+   for (bsoncxx::array::element ele : bson_array) {
+      fc::variant v;
+      from_bson(ele, v);
+      a.push_back(v);
+   }
+}
+
+template <typename T>
+void from_bson(const T& ele, fc::variant& v)
+{
+   switch (ele.type()) {
+      case bsoncxx::type::k_double:
+         v = ele.get_double().value;
+         return;
+      case bsoncxx::type::k_utf8:
+         v = bsoncxx::string::to_string(ele.get_utf8().value);
+         return;
+      case bsoncxx::type::k_document: {
+         fc::mutable_variant_object o;
+         from_bson(ele.get_document().value, o);
+         v = o;
+         return;
+      }
+      case bsoncxx::type::k_array: {
+         bsoncxx::array::view sub_array{ele.get_array().value};
+         fc::variants a;
+         from_bson(sub_array, a);
+         v = a;
+         return;
+      }
+      case bsoncxx::type::k_binary: {
+         fc::blob blob;
+         blob.data.resize(ele.get_binary().size);
+         std::copy(ele.get_binary().bytes, ele.get_binary().bytes + ele.get_binary().size, blob.data.begin());
+         v = blob;
```

Comment:
Have you verified this produces the correct blob value?

Reply:
Yes, but not verified by the chain data. These test codes will produce a  b_binary:

```
  bsoncxx::types::b_binary bin;
  bin.sub_type = bsoncxx::binary_sub_type::k_binary;
  bin.size = 32;
  bin.bytes = reinterpret_cast<const uint8_t *>("\x01\x23\x45\x67\x89\x12\x34\x56\x78\x90\x01\x23\x45\x67\x89\x12\x34\x56\x78\x90\x01\x23\x45\x67\x89\x12\x34\x56\x78\x90\xAB\xCD");
  bsoncxx::builder::stream::document stream_doc{};
  stream_doc << "hex_value" << bin;
  std::cout << bsoncxx::to_json(stream_doc) << std::endl;
```